Map Block

jeffersonrabb opened the pull request:
#### Changes proposed in this Pull Request

* Add Map block to repo, for inclusion in Jetpack.

#### Preview
Screen capture: https://cloudup.com/ceQ6jKr5-R4
Atavist block example (this block is an adaptation of the Atavist version): https://wpmigration.atavist.com/maps-on-atavist

#### Testing instructions

##### Jurassic Ninja

- Click this link: https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=atavist/map-block-v2&branch=try/google-maps-api-key
- Set Up Jetpack/Free Tier
- Activate Recommended Features

##### Jetpack/Docker

- Setup and run the [Jetpack docker environment](https://github.com/Automattic/jetpack/tree/master/docker#docker-environment-for-jetpack-development).
- In Jetpack repo,  checkout `try/google-maps-api-key` branch. 
- In `wp-calypso` repo, checkout [`atavist/map-block-v2`](https://github.com/Automattic/wp-calypso/tree/atavist/map-block-v2) branch.
- Run the following from the root of `wp-calypso` repo`: ```npm run sdk gutenberg client/gutenberg/extensions/presets/jetpack -- --output-dir=PATH_TO_JETPACK_REPO/jetpack/_inc/blocks```  ( replace PATH_TO_JETPACK_REPO with the path to your local Jetpack repo).

##### Vagrant (No longer valid following changes to Google Maps API Key handling)

#### Some questions for discussion

~~1. The block uses browser Geolocation, if available. If no map markers are defined, the block will attempt to use the user's current location. This means that browser will ask for permission to access location, and it significantly slows the initial load. Is this a worthwhile feature? Or does the permission request seem invasive and the delay annoying? If we do opt to keep the feature, at minimum we'll need a loading screen.~~
~~2. How should we handle the Google Maps API key? Currently an API key is hardcoded in the config file, but this is not a sustainable solution. Google requires that a valid key be used in all requests, which must be tied to a billing account. ~~
3. The library used to dynamically load the Google Maps script is [`loadScript`](https://github.com/Automattic/wp-calypso/tree/master/client/lib/load-script). Because of [this issue](https://github.com/Automattic/wp-calypso/issues/27821), importing the library causes errors so as a temporary fix I duplicated the library and placed it in the block directory [here](https://github.com/Automattic/wp-calypso/tree/atavist/map-block-v2/client/gutenberg/extensions/map-block/load-script). After the issue is resolved, I should be able to remove the duplicate and import the original library.
5. The sidebar includes the `Locations` component which allows editing of title/caption for all defined markers. The same editing is possible by clicking on markers on the map, which is the more "Gutenberg-esque" approach. I plan to remove the Locations component, unless people feel that both are necessary.
6. The Map Theme Picker uses images of the themes. SDK builds fail if an image is included (issue [here](https://github.com/Automattic/wp-calypso/issues/27838)) so I am temporarily accessing those images from a public S3 bucket.
7. Any additional map styles we should include?
8. Should a future iteration of the block support other mapping services?

MichaelArestad commented:
This is awesome!

As we discussed earlier, there are some questions about the caption:

1. Atavist has some custom styling for captions. Do we need the caption styles to match the Atavist Map caption?
2. If not, do we need a caption at all? 

On other Gutenberg blocks, we have been allowing folks to add their own blocks to label or caption elements. For example, the subscription form will just be an email input and submit button. We are relying on folks to add a paragraph or heading encouraging folks to use the form. I think we could do something similar here by removing the caption which would slightly simplify this block.

However, if we need it for parity, maybe we could have an "Atavist site" mode and a default mode. In the Atavist mode, we show the caption/controls. In the default mode we don't. I am wary of this because it might be easy to forget to support down the road.

There is also a minor padding issue on the marker where the address collides with the arrow. Easy peasy to fix with a little padding:
![image](https://user-images.githubusercontent.com/1123119/46898724-dfa68700-ce48-11e8-8568-fb280041ff36.png)

I also noticed the input for the address (with autocomplete) is `contenteditable`. I'm not sure what the best move here is as it was easy enough to paste in web content into the field. I'll leave judgement to Luna on this.

I also think we could tweak the color palette of the pins. I'll see what I can do there. I think our brand team has some color palettes we could probably pull from.

jeffersonrabb commented:
@MichaelArestad, a few notes on your comments:

> I also think we could tweak the color palette of the pins.
I changed the color palettes to use Theme colors, matching the Paragraph block background and others. https://github.com/Automattic/wp-calypso/pull/27836/commits/4af0541402ee4f644014bdeed3d417d356fc1efa

> There is also a minor padding issue on the marker where the address collides with the arrow. 
Fixed in https://github.com/Automattic/wp-calypso/pull/27836/commits/67760465caa0158a9ba722c47cd7626239013a8f

On 10/12 @MichaelArestad found out a few issues that caused map markers to stop responding to clicks. These included multiple map blocks in one page, removing and re-adding map block without page refresh, opening a marker and closing the info window by clicking the X. These are now resolved in https://github.com/Automattic/wp-calypso/pull/27836/commits/4d91781bcba6979c33a9ac3fc36770babecd6572

> do we need a caption at all
It looks like all of the media blocks (image|video|audio) have captions, but no others do. Atavist Maps do have a caption, which is styled differently, so in the interest of aiming for full visual parity I'll need some way to do this. Here are the options I can think of:

1) Create "official" Map Block without caption, use Gutenberg [hook & filter extensibility](https://wordpress.org/gutenberg/handbook/extensibility/extending-blocks/) options to append a caption for Atavist migration customers.
2) Include caption in the Map Block, but have it hidden by default. Add sidebar toggle to switch it on/off.
3) Remove caption from Map block, add caption content as styled paragraph blocks in Atavist migration.
4) Consider Map to be more like Image/Video/Audio, include caption as is current state.

What do you think?

jeffersonrabb commented:
> I also noticed the input for the address (with autocomplete) is contenteditable. I'm not sure what the best move here is as it was easy enough to paste in web content into the field.

Gutenberg `Autocomplete` is really designed to work with a `contenteditable` element, not a plain text input. In the first version of this PR I had a hacky solution working, but as @MichaelArestad points out its kind of gross to have a `contenteditable` element masquerading as a text input. I ended up writing a new component, [`Lookup`](https://github.com/Automattic/wp-calypso/blob/042b55bd8e61829a4185ee0d03cb5ce95a2a0a5e/client/gutenberg/extensions/map-block/lookup/index.js), which is based on `Autocomplete` as well as @ryelle's work [here](https://href.li/?https://github.com/woocommerce/wc-admin/pull/322). `Lookup` differs from `Autocomplete` in the following ways:

- Designed to be used with a text input instead of a `contenteditable` div.
- Takes only one `completer` param (for a text input there is only need for one, while a Rich Text element can use many)
- No TriggerPrefixes at all, since anything typed in the input should be immediately checked for a match.
- No node-level assessment and disabling, since the input is a simple text field.

As of https://github.com/Automattic/wp-calypso/pull/27836/commits/042b55bd8e61829a4185ee0d03cb5ce95a2a0a5e `Lookup` is functional, although I still need to add key command and spoken message support to match `Autocomplete`. At a later stage I might propose including it as a core Gutenberg component since I expect this is a use case which will come up again. cc: @mtias, @ryelle

MichaelArestad commented:
> Create "official" Map Block without caption, use Gutenberg hook & filter extensibility options to append a caption for Atavist migration customers

This seems like the cleanest approach. I wonder if the caption could even be a separate block that's hidden for most users?

jeffersonrabb commented:
Caption removed in https://github.com/Automattic/wp-calypso/pull/27836/commits/482b2718143d2ea6daf1ace153b6d1f459510677. Note that if you have a pre-existing Map block you'll get errors and need to remove and replace. Since the block hasn't been released yet, seems like overkill to set up deprecation for the caption.

georgestephanis commented on client/gutenberg/extensions/map-block/edit.js:
```diff
@@ -0,0 +1,146 @@
+ * External dependencies
+ */
+
+import { clone } from 'lodash';
```
Just to have one fewer dependency, what are your thoughts on instead doing something like `let clonedArray = originalArray.slice(0);` which I believe should have the same effect?

https://davidwalsh.name/javascript-clone-array

jeffersonrabb replied:
Good call. Fixed in https://github.com/Automattic/wp-calypso/pull/27836/commits/3bfbe0c922c1ffd2c183aa2bc85a2922a71a1154

tyxla commented on client/gutenberg/extensions/map-block/add-point/index.js:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Wordpress dependencies
```
Seems like those are external dependencies for Calypso. 

Also, mind the capital `P` in `WordPress` - WP core has [a function](https://codex.wordpress.org/Function_Reference/capital_P_dangit) to autocorrect this, but we don't :)

This applies to other occurences in this PR.

jeffersonrabb replied:
Oops, yes will globally fix all lower p WordPress. Re: the external dependencies note are you suggesting I should remove the `Wordpress dependencies` comment block everywhere and simply move those imports to `External dependencies`?

tyxla replied:
Moving them all to `External dependencies` and removing the `WordPress` one makes sense 👍

jeffersonrabb replied:
Great, will do! I did pick up the pattern from other JetPack block code, e.g.: https://github.com/Automattic/wp-calypso/blob/master/client/gutenberg/extensions/tiled-gallery/editor.js#L3-L5. Would we want to make the same change everywhere?

simison replied:
Yep, I've done the change in the Tiled gallery as well 👍 (pending merge) and good to PR changes anywhere else you spot it.

The pattern came from Gutenberg core blocks originally.

jeffersonrabb replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/27836/commits/a093d10a6ce2ad3116c67022db1ac1d94300d5ce

tyxla commented on client/gutenberg/extensions/map-block/add-point/index.js:
```diff
@@ -0,0 +1,73 @@
+ */
+
+import { __ } from '@wordpress/i18n';
+
```
Having empty lines between imports of the same type leaves a lot of noise. What do you think about removing these empty lines where possible? If in doubts, just have a look at some of the client code in Calypso.

jeffersonrabb replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/27836/commits/a093d10a6ce2ad3116c67022db1ac1d94300d5ce

tyxla commented on client/gutenberg/extensions/map-block/add-point/index.js:
```diff
@@ -0,0 +1,73 @@
+
+	constructor() {
+		super( ...arguments );
+		this.onAddPoint = this.onAddPoint.bind( this );
```
If we make `onAddPoint` an arrow function, it will be automatically binded and we won't need a `constructor` at all.

Anyway, `this.onAddPoint` seems to be unused, so we can directly remove the `constructor` too.

simison replied:
BTW this is clearly a pattern coming from Gutenberg core blocks.

jeffersonrabb replied:
Good idea, I prefer the arrow functions too - replaced all of them in https://github.com/Automattic/wp-calypso/pull/27836/commits/92aa832fd43ffac9ffdbc63ec43c68a9d9346760

tyxla commented on client/gutenberg/extensions/map-block/add-point/index.js:
```diff
@@ -0,0 +1,73 @@
+
+	onAddPoint( point ) {
+
+		this.props.onAddPoint( point );
```
In cases like this, we could directly call `this.props.onAddPoint` without the need to define a wrapper method.

But also, this method doesn't seem to be used at this time, so we can directly remove it.

jeffersonrabb replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/27836/commits/92aa832fd43ffac9ffdbc63ec43c68a9d9346760

tyxla commented on client/gutenberg/extensions/map-block/add-point/index.js:
```diff
@@ -0,0 +1,73 @@
+			<Button
+				className='map__add_btn'
+			>
+				Add point
```
Let's make sure to localize all the copy.

jeffersonrabb replied:
Got this in https://github.com/Automattic/wp-calypso/commit/11702a5987ce8bc9503ddc5eb0a180ee50fe9aec

tyxla commented on client/gutenberg/extensions/map-block/add-point/index.js:
```diff
@@ -0,0 +1,73 @@
+		} = this.props;
+		return (
+			<Button
+				className='map__add_btn'
```
We usually use double quotes for string literals that we pass as props.

jeffersonrabb replied:
Not sure I follow... where is a string literal being passed as props here?

tyxla replied:
Oh, well `'map__add_btn'` is a literal of type string. Note that I mean string literal, not template literal (those 2 are often confused with each other).

jeffersonrabb replied:
This was resolved (along with many other syntax issues) in https://github.com/Automattic/wp-calypso/commit/bca3b6c11f2bca13d7a57d781e8cc40f1c0810b1 after running all files through `JsPrettier`.

tyxla commented on client/gutenberg/extensions/map-block/add-point/index.js:
```diff
@@ -0,0 +1,73 @@
+}
+
+AddPoint.defaultProps = {
+	onAddPoint: () => {},
```
Use lodash's `noop` instead of defining empty functions?

jeffersonrabb replied:
I'm happy to, but two notes on why I did this. 1) I pulled the pattern from https://github.com/WordPress/gutenberg/blob/master/packages/components/src/form-token-field/index.js#L601. 2) In an [earlier comment](https://github.com/Automattic/wp-calypso/pull/27836#discussion_r226453721) in this PR, @georgestephanis had suggested I take steps to remove `lodash` dependencies, but this would be adding one. Given these, do you still like this change? I'm more than happy to go either way, just asking about consistency.

tyxla replied:
Ah, if our goal is to reduce `lodash` dependencies, then let's leave it as-is 👍

tyxla commented on client/gutenberg/extensions/map-block/add-point/oval.svg:
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This SVG can be cleaned up and minified.

jeffersonrabb replied:
Since the raster assets are going to be in the WP.com CDN, I opted to move this one too in https://github.com/Automattic/wp-calypso/pull/27836/commits/c778a244c2702e3e96b782aa062d8c96e243f57c. Currently it is in the Atavist S3 but will be moved to WP.com shortly.

tyxla commented on client/gutenberg/extensions/map-block/add-point/style.scss:
```diff
@@ -0,0 +1,50 @@
+.map__add_btn {
```
Seems like some classes don't follow Calypso's coding guidelines: https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#general-syntax-and-writing-rules

Particularly, this one should be `.map__add-btn` instead.

Also, I'd suggest using our automatically generated class that's coming from `this.props.className` - it allows us to namespace our styles a bit better.

tyxla commented on client/gutenberg/extensions/map-block/config.js:
```diff
@@ -0,0 +1,338 @@
+ */
+
+export const CONFIG = {
+	name: 'atavist/maps',
```
Probably makes sense to use the same prefix like our other blocks here - `a8c`.

jeffersonrabb replied:
Just to be sure I understand, you're suggesting replacing `atavist/` with `a8c/`? I would probably also rename the branches too, e.g. `a8c/map-block-v2`? I assume this change will break all existing instances of the block (purely testing, so no big deal) but I want to do this all at once if I'm going to make that kind of change.

jeffersonrabb replied:
Change made in https://github.com/Automattic/wp-calypso/pull/27836/commits/be1654f671aa25df8f79a332b3dda5f2ff4c7169, going to leave the branch name intact for now.

tyxla commented on client/gutenberg/extensions/map-block/config.js:
```diff
@@ -0,0 +1,338 @@
+
+export const CONFIG = {
+	name: 'atavist/maps',
+	title: __( 'Map' ),
```
If this will be a Jetpack block, does it make sense to have the `jetpack` textdomain? What are our plans on handling and serving translations for these blocks?

jeffersonrabb replied:
Not sure I understand this one - could you give me some context?

tyxla replied:
Well, in order for `wp.i18n` to know where to look for translations for a block, it needs a textdomain. In `wp.i18n.__()`, this is the second argument of the function. In this particular call, we seem to be lacking a textdomain.

If you have a look ot some of the other Jetpack blocks, we use the `jetpack` textdomain, which will later allow us to use Jetpack for indexing translations, then managing and serving them.

Let me know if that makes sense.

jeffersonrabb replied:
Yes, thank you for the explanation. I believe I got them all in https://github.com/Automattic/wp-calypso/commit/0241560c0f84181a9c2f4fedfe30f90ca084d0f4. Could you take a look and verify I did this correctly?

tyxla replied:
All the `__()` cases look good to me 👍 The only case that seem to have been missed is the `_n()` one - https://github.com/Automattic/wp-calypso/pull/27836/files#diff-7e7cc0bd139c6be3b8dcca033daea924R169 - it needs the textdomain as a fourth argument.

jeffersonrabb replied:
Caught the `_n()` case in https://github.com/Automattic/wp-calypso/pull/27836/commits/bf7f6a08e4fe3e5f4314c556bf6458c62eaa6a1e.

tyxla commented on client/gutenberg/extensions/map-block/config.js:
```diff
@@ -0,0 +1,338 @@
+	icon: <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M20.5 3l-.16.03L15 5.1 9 3 3.36 4.9c-.21.07-.36.25-.36.48V20.5c0 .28.22.5.5.5l.16-.03L9 18.9l6 2.1 5.64-1.9c.21-.07.36-.25.36-.48V3.5c0-.28-.22-.5-.5-.5zM15 19l-6-2.11V5l6 2.11V19z"/><path d="M0 0h24v24H0z" fill="none"/></svg>,
+	category: 'common',
+	keywords: [
+		__( 'Map' ),
```
I think it's very common for blocks to have keywords in lowercase.

jeffersonrabb replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/27836/commits/4dd86609b220eba1c4f2c08f611ff97809c612b9. Also, added "Jetpack" as a keyword.

tyxla commented on client/gutenberg/extensions/map-block/config.js:
```diff
@@ -0,0 +1,338 @@
+ * Internal dependencies
+ */
+
+export const CONFIG = {
```
If we want to be consistent with how Gutenberg core do things, we could name this `settings`:

https://github.com/WordPress/gutenberg/blob/e7d674dc5ee245a41a8c2e9fac88a021dbaded32/packages/block-library/src/audio/index.js#L16

jeffersonrabb replied:
Done: https://github.com/Automattic/wp-calypso/pull/27836/commits/636f5c882802e7d50f02eeb9d32f8bf46bade81d

tyxla commented on client/gutenberg/extensions/map-block/edit.js:
```diff
@@ -0,0 +1,144 @@
+			addPointVisibility: false,
+		};
+		this.mapRef = createRef();
+		this.addPoint = this.addPoint.bind( this );
```
Seems like we could remove all the `bind` calls here if we use arrow functions.

jeffersonrabb replied:
Done https://github.com/Automattic/wp-calypso/pull/27836/commits/92aa832fd43ffac9ffdbc63ec43c68a9d9346760

tyxla commented on client/gutenberg/extensions/map-block/info-window/index.js:
```diff
@@ -0,0 +1,86 @@
+
+export class InfoWindow extends Component {
+
+	constructor() {
```
This method is also removable if we make `closeClick` an arrow function.

jeffersonrabb replied:
Done in https://github.com/Automattic/wp-calypso/pull/27836/commits/92aa832fd43ffac9ffdbc63ec43c68a9d9346760

tyxla commented on client/gutenberg/extensions/map-block/info-window/index.js:
```diff
@@ -0,0 +1,86 @@
+		this.el = document.createElement( 'DIV' );
+		this.infowindow = new google.maps.InfoWindow( {
+      		content: this.el
+    	} );
```
There seem to be indentation issues in this file.

jeffersonrabb replied:
Couldn't see this one either?

jeffersonrabb replied:
Fixed in https://github.com/Automattic/wp-calypso/pull/27836/commits/1929d54755a21d7ff44905f66f55fb800ba2723d

tyxla commented on client/gutenberg/extensions/map-block/info-window/index.js:
```diff
@@ -0,0 +1,86 @@
+	componentDidMount() {
+
+		const { google } = this.props;
+		this.el = document.createElement( 'DIV' );
```
Could we avoid any direct DOM manipulations inside components?

jeffersonrabb replied:
This is a tricky one. When a map marker is clicked, I open a [Google Maps Info Window](https://developers.google.com/maps/documentation/javascript/infowindows). When you do this, you can pass in plain text, markup, or a DOM node. Because I'm allowing editing of the fields within the Info Window, I need to be able to render React components into it, which I do by creating this DIV and then treating is as a React Portal. So even though that `createElement` looks suspicious, its actually setting up a React pattern - a portal.

tyxla commented on client/gutenberg/extensions/map-block/load-script/index.js:
```diff
@@ -0,0 +1,65 @@
+	}
+}
+
+export function loadjQueryDependentScript( url, callback ) {
```
I'm curious why we need this - it doesn't seem to be used anywhere.

jeffersonrabb replied:
Quick note about this, and all of `load-script`... This is a Calypso library found here: https://github.com/Automattic/wp-calypso/tree/master/client/lib/load-script. I want to directly import the original, but doing so leads to a breaking error that I filed here: https://github.com/Automattic/wp-calypso/issues/27821. For a short term fix I copied the library into the block directory, and hackily ditched the lines that were causing the error, which unblocked me. As soon as the issue is resolved, I'll remove the duplicated library and change the import. So the short answer here is, there's not much point making any changes to `load-script` since it's not long for this world :-)

tyxla commented on client/gutenberg/extensions/map-block/map-theme-picker/style.scss:
```diff
@@ -0,0 +1,38 @@
+		border-color: black;
+	}
+	&.default {
+		background-image: url( 'https://s3.amazonaws.com/atavist-static/map-themes/map-theme-default.jpg' );
```
Do we want to host these images in WP.com?

jeffersonrabb replied:
Originally I wanted to include these in the block directory, but I ran into this issue: https://github.com/Automattic/wp-calypso/issues/27838. Even if resolved, your question is still quite valid - is it good practice to include assets in source code or should they just go on a CDN. And if CDN is the answer, we should definitely move them from this old Atavist one. Where would I put them on WP.com?

simison replied:
I'm not sure if we can use Google map layer image without valid API code in place first. If we can, The image should probably have Google -logo+copyrights at the corner.

Shipping the image with the block feels like the way to go but the SDK might not support it yet. 🤔  WP.com CDN sounds like a sensible alternative meanwhile.

jeffersonrabb replied:
Ah, this is a mistake, left in from an ancient Atavist version. I'll remove!

jeffersonrabb replied:
Actually, my earlier comment was wrong - these images are indeed meant to be there. These images are used in the circles to demonstrate the different style options:

<img width="278" alt="screen shot 2018-10-19 at 9 54 41 am" src="https://user-images.githubusercontent.com/1477002/47222556-10c11300-d385-11e8-913f-ac70a286afc2.png">
 
The WP.com CDN solution sounds like the right approach, but @simison can you say a little more about what you meant by this?

> I'm not sure if we can use Google map layer image without valid API code in place first. If we can, the image should probably have Google -logo+copyrights at the corner.

simison replied:
I meant that typically Google isn't very happy if their map tiles are used without logo/copyright attribution visible somewhere, but not sure if it matters in this case. I'm sure it's buried somewhere in their [terms](https://cloud.google.com/maps-platform/terms/).

Could these images be loaded directly from Google servers via [Maps Static API](https://developers.google.com/maps/documentation/maps-static/intro), using the same API key that user provides?

The theme selection could stay hidden until the API key is provided.

sirreal replied:
Related PR to include image assets (pngs): #28378

jeffersonrabb replied:
This question resolved and the CDN question superseded by the work in https://github.com/Automattic/wp-calypso/pull/28378 (thank you @sirreal!) Moved the images into the bundle in https://github.com/Automattic/wp-calypso/pull/27836/commits/ea07f2e853409425716133aed494a53afd9d2107

tyxla commented on client/gutenberg/extensions/plugins/atavist/plugin.php:
```diff
@@ -0,0 +1,49 @@
+<?php
```
Ouch, a PHP plugin in Calypso. Doesn't seem to be the right place for this - if this was going in Jetpack, I'd suggest  an accompanying PR in Jetpack instead. Then we wouldn't be needing any of it if we use the Jetpack blocks preset.

simison replied:
This feels still an open question.

If the Atavist block should always depend on Jetpack, then it makes sense to move PHP to Jetpack as well. If it needs to work with both Jetpack and Atavist somewhere else without Jetpack, it gets more complicated.

cc @dmsnell

jeffersonrabb replied:
Discussion here: p1537557809000100-slack-jetpack-gutenberg

dmsnell replied:
yeah 🤷‍♂️ we're playing around with it. there's no reason PHP can't live in Calypso's repo. I think it's worthwhile exploring but I don't have an answer.

tyxla commented on client/gutenberg/extensions/shared/atavist/frontend-management.js:
```diff
@@ -0,0 +1,78 @@
+	}
+
+	extractChildrenFromContainer( node ) {
+    	const children = [];
```
There seem to be indentation issues here.

jeffersonrabb replied:
🤔At first glance, I'm not seeing any—this file looks to be all tabbed out. Specifically what did you notice?

dmsnell replied:
line 60 has a tab but line 61 has spaces

jeffersonrabb replied:
Right you are! Fixed in https://github.com/Automattic/wp-calypso/pull/27836/commits/1929d54755a21d7ff44905f66f55fb800ba2723d. How did you spot these?

sirreal commented:
This will require https://github.com/Automattic/jetpack/pull/10362 or similar for locales to work properly on the frontend.

enejb commented on client/gutenberg/extensions/map-block/add-point/style.scss:
```diff
@@ -0,0 +1,53 @@
+	height: 38px;
+	margin-top: -19px;
+	margin-left: -16px;
+	background-image: url( https://s3.amazonaws.com/atavist-static/map/oval.svg );
```
Just a note that we will need to move this file to be loaded from the Jetpack side directly.

jeffersonrabb replied:
There was discussion of moving all the static assets to WP.com ( https://github.com/Automattic/wp-calypso/pull/27836#discussion_r226570934 ).

jeffersonrabb replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/27836/commits/ea07f2e853409425716133aed494a53afd9d2107

enejb commented on client/gutenberg/extensions/map-block/editor.scss:
```diff
@@ -0,0 +1,11 @@
+/** @format */
+
+.wp-block-atavist-maps {
```
Do we need these empty styles?

enejb commented on client/gutenberg/extensions/map-block/save.js:
```diff
@@ -0,0 +1,46 @@
+				case 'right':
+				case 'center':
+				case 'full':
+					return 'atavist-block-align-' + value;
```
Can we rename `atavist-block-align-` to `jetpack-block-align-` ?

jeffersonrabb replied:
Yes, will do. I will need the `atavist-`* alignment classes in order for this block to play nicely with Atavist themes, but I believe I'll be able to accomplish this using Gutenberg's extensibility hooks, so I'll put these back in through a plugin.

jeffersonrabb replied:
Got rid of all of the atavist class names in https://github.com/Automattic/wp-calypso/pull/27836/commits/be1654f671aa25df8f79a332b3dda5f2ff4c7169.

enejb commented on client/gutenberg/extensions/plugins/atavist/plugin.php:
```diff
@@ -0,0 +1,49 @@
+<?php
```
Is this code necessery? I though that Jetpack would determine when this code gets loaded.

jeffersonrabb replied:
This file is actually unrelated to Jetpack. The Map block is going to be included in Jetpack but this (and other similar blocks) are also being included in a stand-alone plugin, and this `.php` file relates to the plugin only.

enejb commented:
I did some testing and it works really well. Nicely done. 

I noticed that the code creates the following block html. 
```
<!-- wp:atavist/maps {"zoom":9} -->
<div class="wp-block-atavist-maps atavist-block atavist-simple-map atavist-block-align-center" data-map_style="default" data-points="[]" data-zoom="9" data-map_center="{&quot;latitude&quot;:37.7749295,&quot;longitude&quot;:-122.41941550000001}" data-marker_color="red" data-api_key="code...."></div>
<!-- /wp:atavist/maps -->
```

I think if the code is being included in Jetpack we should prefix it with Jetpack instead of atavist. This way the user would know that it they would need to reactivate Jetpack to get the component to work again. I am also wonder if we could use some sort google maps iframe or image tag as the default view instead of a div. This way the map block would should something to the user even in environments where js is not supported. Such as when the content appears inside a feed reader, a subscription email or when the plugin is deactivated. 

When it comes to the google maps api key on the .com side we have a few different filters that you can use that will return the expected api key. We currently do not expose thats key in Jetpack even for the paid users. 

They are required to enter it as part of the widget UI.
See: 
<img width="441" alt="screen shot 2018-10-22 at 4 29 48 pm" src="https://user-images.githubusercontent.com/115071/47325223-c0adaf00-d617-11e8-8202-90ede1e9ec6b.png">

What the plan is for the other stand alone plugin? Where can I read more about it?

dmsnell commented on client/gutenberg/extensions/map-block/edit.js:
```diff
@@ -125,7 +219,15 @@ class MapEdit extends Component {
+				{ noticeUI }
+				{ apiState === API_STATE_LOADING && placholderAPIStateLoading }
+				{ apiState === API_STATE_FAILURE && placeholderAPIStateFailure }
+				{ apiState === API_STATE_SUCCESS && placeholderAPIStateSuccess }
```
nice state-machine-based approach 😉

dmsnell commented on client/gutenberg/extensions/map-block/edit.js:
```diff
@@ -45,10 +66,46 @@ class MapEdit extends Component {
+				this.setState( {
+					apiState: API_STATE_FAILURE,
+				} );
+			} );
```
this is fine here I think given the code that's actually executing but a quick note on `.then()` is that we can pass two parameters or one. with two parameters the second is actually the `catch` but it differs from `.then().catch()` because when both are supplied to `then` then the catch only catches from the original promise.

I'm sure that's confusing, so here is some code

```js
Promise
	.resolve()
	.then( () => { throw 'a' }, () => { throw 'b' } )
	// threw 'a'

Promise
	.reject()
	.then( () => { throw 'a' }, () => { throw 'b' } )
	// threw 'b'

Promise
	.resolve()
	.then( () => { throw 'a' } )
	.catch( () => { throw 'b' } )
	// threw 'b'

Promise
	.reject()
	.then( () => { throw 'a' } )
	.catch( () => { throw 'b' } )
	// threw 'b'
```

notice how in the last two cases `'b'` threw in both the resolution and rejection part? that's because _the exception which happened inside of the `then()` cascaded into the following `catch()` chain_. it _looks_ like the `catch()` is there to grab the failure from the original promise but actually it's grabbing _both_ a failure in the original promise _and_ a failure in the chained `then()`

hope that's useful! (@blowery is good at finding bugs from this type of thing and fixing them)

enejb commented on client/gutenberg/extensions/map-block/settings.js:
```diff
@@ -0,0 +1,406 @@
+		</svg>
+	),
+	category: 'common',
+	keywords: [ __( 'map', 'jetpack' ), __( 'jetpack', 'jetpack' ), __( 'jetpack' ) ],
```
` __( 'jetpack', 'jetpack' ), __( 'jetpack' ) `
I am not sure we need to translate `jetpack`. Since it is a product name. 
might be good to add `location` and `google` to this list.

enejb commented on client/gutenberg/extensions/map-block/settings.js:
```diff
@@ -0,0 +1,406 @@
+			],
+		},
+		terrain_with_features: {
+			map_type: 'TERRAIN',
```
Map type is the same as above.

enejb replied:
My guess is this is intentional. Just wanted to bring it up in case it wasn't. The same goes for the map_type below.

jeffersonrabb replied:
Yes, this is intentional. The map type is the same, but the difference is in the labels and a couple of other parameters.

enejb commented:
I test a bunch of testing of this PR and found a couple of issues. 

Issue 1.
When I click on the alignment buttons. The Map goes away. 
![screen_shot_2018-10-25_at_4_18_06_pm](https://user-images.githubusercontent.com/115071/47535923-75dba380-d872-11e8-988b-4f2c28ae6b99.png)

and get 
![screen shot 2018-10-25 at 4 24 01 pm](https://user-images.githubusercontent.com/115071/47535932-7b38ee00-d872-11e8-8ef7-eb0bb585b4b8.png)

Issue 2.
Also for some reason the map doesn't show up for me on the front page any more.
Instead I see the following
![screen shot 2018-10-25 at 4 25 23 pm](https://user-images.githubusercontent.com/115071/47535962-9a378000-d872-11e8-8476-f855da69a8f5.png)

Here is the URL of the page https://motionless-seals-368.jurassic.ninja/this-is-fun/ 

From the UI point of view I think it would be nice if we just had 4 different map options (terrain, satellite view, simple view and black and white ) and a toggle that lets us add the more detailed location info. (city, streets etc... ) Then we would end up with a total of 8 different variation but the user would only be choosing between 4 options :)

jeffersonrabb commented:
> When I click on the alignment buttons. The Map goes away.

This is fixable, however after considering this a bit the map block doesn't work terribly well floated left or right. The block generally looks much better centered. So, I resolved this issue by removing the left/right alignment options (in https://github.com/Automattic/wp-calypso/pull/27836/commits/1c47ada219ec25ac2883a602da748e2181be60c9). Curious if @MichaelArestad or @kjellr think that removing the floated alignments is a mistake?

jeffersonrabb commented:
> Also for some reason the map doesn't show up for me on the front page any more.

This was an error (mismatch with changes made in the API), which is fixed in https://github.com/Automattic/wp-calypso/pull/27836/commits/7ce9855bfd071e1d9f64cc93314c6ca707f2ad99. However, there are legitimate situations that could cause the frontend map not to load, such as removal of valid Google API key without replacement, API key is deleted in Google Console, etc. Because of this we should give some consideration to what the frontend version looks like if the map cannot render—currently as shown here it will just be an empty gray rectangle which is no good. @MichaelArestad could you weigh in on this?

jeffersonrabb commented:
> From the UI point of view I think it would be nice if we just had 4 different map options (terrain, satellite view, simple view and black and white ) and a toggle that lets us add the more detailed location info. (city, streets etc... ) Then we would end up with a total of 8 different variation but the user would only be choosing between 4 options :)

@MichaelArestad, @kjellr what do you think about this suggestion?

kjellr commented:
> From the UI point of view I think it would be nice if we just had 4 different map options (terrain, satellite view, simple view and black and white ) and a toggle that lets us add the more detailed location info. (city, streets etc... ) Then we would end up with a total of 8 different variation but the user would only be choosing between 4 options :)

That sounds reasonable to me!

jeffersonrabb commented:
> From the UI point of view I think it would be nice if we just had 4 different map options (terrain, satellite view, simple view and black and white ) and a toggle that lets us add the more detailed location info. (city, streets etc... ) Then we would end up with a total of 8 different variation but the user would only be choosing between 4 options :)

I gave this a shot - neat idea! Try out here: [JN](https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=atavist/map-block-themepicker-v2&branch=try/google-maps-api-key)

kjellr commented:
Oh also:

> Curious if @MichaelArestad or @kjellr think that removing the floated alignments is a mistake?

This seems okay by me. I would like to see wide/full alignments for this block though.

enejb commented:
Nice, I really like that we removed the left/right alignment and reducing the number of options. With the added ability to toggle the street info. 

One thing that I notice is that when a user removes the API key (by clicking the button). There is no warning. I think we should tell the user that they are about to break all the map blocks on their site and getting them to confirm that choice. 

When the user fist sees the map. They encounter the map of SF. Which I think is good default but it could be a bit confusing to get to the location that they really want. Since they might not realize that the default is SF. 

I wonder if the user dismisses the adding a new marker UI. If we can show them some sort of input where they can enter the address that helps them center the map. cc: @MichaelArestad and @kjellr can explore some designs for that? 

Another thing that I noticed is that we show the user the button to Take the map full screen. Which I think is definitely handy to have on the page view but could be a bit strange if clicked inside the editor. I can see arguments for keeping it as it is. I just wanted to bring it for discussion. :)

jeffersonrabb commented:
> I would like to see wide/full alignments for this block though.

@kjellr the block does support wide/full alignments already, but this is contingent on theme support. If you try it out with twentynineteen, for example, you'll have those alignments available as shown here:

<img width="1645" alt="screen shot 2018-10-26 at 12 19 35 pm" src="https://user-images.githubusercontent.com/1477002/47579328-a1659900-d919-11e8-91d5-b123f0e6d333.png">

jeffersonrabb commented:
> One thing that I notice is that when a user removes the API key (by clicking the button). There is no warning. I think we should tell the user that they are about to break all the map blocks on their site and getting them to confirm that choice.

I was actually planning to get rid of the REMOVE KEY functionality entirely. It's just there to enable testing, so that we can easily replicate what the block feels like to a first-time user without needing to spin up a new JN instance or delete the key in the database. There isn't much of a good reason to remove a key that I can think of. Either you replace with another valid key, or leave it be. Can you think of a scenario where deleting the key is needed?

enejb commented:
I do think that for most users trying to show their location on a map. The current flow is great! But if you just want to show a map there is an issue. Although I can't think of a good reason to just show the just the map without a location marker. 

I think it would be good to provide a way to center just the map over a particular location. Without adding a marker. The current way that it is done you have to drag the map and you have to zoom out and in to the desired location. Or you add a marker and then then delete it which feels a bit backwards. If you don't want to show a marker on your map. 

I don't consider this a blocker but more of an improvement to the current flow that could be done later. 

> I was actually planning to get rid of the REMOVE KEY functionality entirely. It's just there to enable testing, so that we can easily replicate what the block feels like to a first-time user without needing to spin up a new JN instance or delete the key in the database. There isn't much of a good reason to remove a key that I can think of. Either you replace with another valid key, or leave it be. Can you think of a scenario where deleting the key is needed?

I can't think of a good reason for the user to remove the key. So just removing the button from the UI works.

jeffersonrabb commented:
> I think it would be good to provide a way to center just the map over a particular location. Without adding a marker. The current way that it is done you have to drag the map and you have to zoom out and in to the desired location. Or you add a marker and then then delete it which feels a bit backwards. If you don't want to show a marker on your map.

One other quick note about this. The original Atavist block actually _does_ have this feature, but I removed it because I thought it added unnecessary complexity since you could achieve nearly the same thing by adding a single pin. The Atavist block allowed a choice between "Fit All Markers" and "Focus on Location." If "Focus on Location" is selected, an additional field is revealed that allows for selection of the central location. I found this confusing, so I opted to remove in favor and simply use "Fit All Markers" in all cases. Here's a screenshot:

<img width="1382" alt="screen shot 2018-10-26 at 12 52 49 pm" src="https://user-images.githubusercontent.com/1477002/47580968-3e2a3580-d91e-11e8-8159-75c4128d3814.png">

MichaelArestad commented:
> Curious if @MichaelArestad or @kjellr think that removing the floated alignments is a mistake?

I agree with Kjell. I do think folks will want to float it left or right to make a "contact info" grouping of some sort, though. It would be nice to have.

> Because of this we should give some consideration to what the frontend version looks like if the map cannot render—currently as shown here it will just be an empty gray rectangle which is no good. @MichaelArestad could you weigh in on this?

I think we can work something up for specific situations where it wont render. I would make it look very similar to the design of it before it has an API key. Perhaps with instructions on how to fix it.

As for the discussion on alternate ways to use the marker, I think we're going to want to see how folks use it in the call for testing. I didn't have the presumption that I needed to move the pin, but perhaps it's as simple as a centered input that isn't attached to a pin. In other words, I think the outline of the pin might be the confusing part.

MichaelArestad commented:
Does a map have to have a marker?

jeffersonrabb commented:
> I agree with Kjell. I do think folks will want to float it left or right to make a "contact info" grouping of some sort, though. It would be nice to have.

To be clear, @kjellr had said he was ok with removing left/right but had asked for wide and full (which are already availabile, assuming the theme supports them). The reason I removed them is that the map starts to look pretty odd at half (or less) size. Could revisit, though.

jeffersonrabb commented:
> I think we can work something up for specific situations where it wont render. I would make it look very similar to the design of it before it has an API key. Perhaps with instructions on how to fix it.

I'm not sure you want to show this kind of UI design on the frontend, because it would be seen by normal unauthenticated viewers. Asking a site visitor to acquire and update a Google API key would confuse them (or worse if they thought they were being asked to pay). Also, introducing actionable UI on the frontend kind of blurs the line between editor and frontend, which would have some tricky technical implications. I would say we should either entirely hide the map on the frontend, or display some sort "This map is unavailable" UI.

jeffersonrabb commented:
> Does a map have to have a marker?

Does not have to, but if it doesn't it will center on the default location (currently SF). This is the topic @enejb was bringing up in earlier comments about a new 'central location' field, like the one in the original Atavist block.

MichaelArestad commented:
> Does not have to, but if it doesn't it will center on the default location (currently SF). This is the topic @enejb was bringing up in earlier comments about a new 'central location' field, like the one in the original Atavist block.

Ah. Yeah. I think this would be prime territory for an Inspector setting in a future PR/iteration.

jeffersonrabb commented:
> Ah. Yeah. I think this would be prime territory for an Inspector setting in a future PR/iteration.

Cool, I'll try this in the next day. My fear is that we introduce confusion by including two separate UI elements for looking up locations, but maybe it won't be so confusing if its tucked away in the sidebar.

MichaelArestad commented:
> Cool, I'll try this in the next day. My fear is that we introduce confusion by including two separate UI elements for looking up locations, but maybe it won't be so confusing if its tucked away in the sidebar.

If you do, do it in a separate PR so it doesn't slow down the merge of this one. If you add it, it should feel like an advanced setting and perhaps have a clear explanation of what it is for.

jeffersonrabb commented:
> Ah. Yeah. I think this would be prime territory for an Inspector setting in a future PR/iteration.

Here's that feature, on a separate branch: https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=atavist/map-block-center-point&branch=try/google-maps-api-key. At the top of the sidebar there's field for selecting a central location. If you populate it, the map centers there. This field is only visible so long as there are no markers defined. As soon as one is defined the field is hidden and the map centering will be based only on the markers. cc: @kjellr, @MichaelArestad, @enejb

enejb commented:
I just tested it out and I like it! Nicely done.

jeffersonrabb commented:
Taking stock of where we are on this guy. Here's a quick list of what I believe to be outstanding before we're ready for testing...

- [ ] New branch with @enejb's idea for an improved map theme picker. Any objects to me merging this in? JN link: https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=atavist/map-block-themepicker-v2&branch=try/google-maps-api-key
- [ ] New branch with option to set a center point if no markers are selected, similar to the original Atavist block and suggested by @enejb. Bring it in? JN link: https://jurassic.ninja/create/?gutenberg&gutenpack&shortlived&jetpack-beta&calypsobranch=atavist/map-block-center-point&branch=try/google-maps-api-key
- [ ] Some accessibility work to be done. 
- [ ] Per a suggestion from @enejb I plan to add an iFramed Google Map as the fallback markup, in case the React component fails for some reason on the frontend.
- [ ] Need instructions for Google Maps API key acquisition.
- [ ] Need to establish UI for the frontend if the Google Maps Key fails for any reason. The easiest solution would be to suppress display entirely, but other possibilities would be a message that the block is not currently available, or perhaps just fall back to the iFrame mentioned above.
- [ ] Some open questions about permissions/access rules in the Jetpack PR: https://github.com/Automattic/jetpack/pull/10393

Am I missing anything? cc: @enejb, @MichaelArestad, @kjellr

kjellr commented:
> New branch with @enejb's idea for an improved map theme picker. Any objections to me merging this in?

This looks great!

>  New branch with option to set a center point if no markers are selected, similar to the original Atavist block and suggested by @enejb. Bring it in? 

This works, but it's a little weird to use. It's weird that you can set this in the sidebar, not the main interface... and it's also weird that the option disappears without warning when you add a point.  

This quick  exploration doesn't look quite right, but maybe having an option like this on the first point is a better solution?

<img width="613" alt="screen shot 2018-10-30 at 9 54 29 am" src="https://user-images.githubusercontent.com/1202812/47722944-e0516280-dc29-11e8-85ca-bd9375ba46f0.png">

simison commented on client/gutenberg/extensions/map-block/component.js:
```diff
@@ -0,0 +1,343 @@
+import { __ } from '@wordpress/i18n';
+import { Component, createRef, Fragment } from '@wordpress/element';
+import { Button, Dashicon, TextareaControl, TextControl } from '@wordpress/components';
+import { get, assign } from 'lodash';
```
Sadly this format of importing `lodash` is currently causing fatal errors in Gutenberg, since it causes us to clash with `underscores` library used by wp-admin Media components.

We should use this format for now instead:
```js
import get from 'lodash/assign';
import get from 'lodash/get';
```

See https://github.com/Automattic/wp-calypso/pull/28157 for another example.

I hope we can change this back to this format soonish.

jeffersonrabb replied:
Yeah, saw that, will change. Does this apply to `lodash` imports exclusively?

jeffersonrabb replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/27836/commits/ee85e9bab8a71a126b62cfb28705df545f46e730 and https://github.com/Automattic/wp-calypso/pull/27836/commits/d2c809e3f45eaf3a76f33ef56df0e6812a46323e. This was why even though the Map block functioned well other core blocks failed (e.g. Image). Initially assumed this was unrelated to this branch, but turns out the `lodash` imports were the culprit. Thank you @simison for pointing this out!

simison commented on client/gutenberg/extensions/map-block/component.js:
```diff
@@ -0,0 +1,343 @@
+		this.setState( { map }, () => {
+			this.sizeMap();
+			this.mapRef.current.addEventListener( 'alignmentChanged', this.sizeMap );
+			window.addEventListener( 'resize', this.sizeMap );
```
This can be pretty aggressive on browsers performance if not throttled. I'd recommend using loadsh denounce, something like this:

`debounce( this.sizeMap, 200 )`

jeffersonrabb replied:
For the window resize event, you mean?

jeffersonrabb replied:
Debounced. https://github.com/Automattic/wp-calypso/pull/27836/commits/0964c39fb81390eb681074037bb001510fd1cb0c

simison commented on client/gutenberg/extensions/map-block/settings.js:
```diff
@@ -0,0 +1,94 @@
+			<path d="M0 0h24v24H0z" fill="none" />
+		</svg>
+	),
+	category: 'common',
```
Beforr shipping with Jetpack preset, let's make sure to have this as 'jetpack' so that it'll appear under jetpack category. We can later look into making this dynamic for Atavist using the SDK.

jeffersonrabb replied:
This was actually just done in https://github.com/Automattic/wp-calypso/pull/27836/commits/3306a87b33357df4ab1b8216829b2cfdc0046252

oskosk commented:
Seems like Mapbox refers to the secret as Access tokens instead of keys. I found it confusing because I just read the placeholder text mildly. Although I copied what seemd to be my *Default Public Token* and the map just worked. 

Should we take this in consideration and update the terms to be consistent with each other ? It's properly referred to as "Access token" on the sidebar, above the text input for the token, but the buttons to remove or edit still refer to it as key.

![image](https://user-images.githubusercontent.com/746152/47939624-c1233100-dec6-11e8-9b0f-c36cb375dd59.png)


![image](https://user-images.githubusercontent.com/746152/47939702-faf43780-dec6-11e8-924c-e4828f28dd2b.png)

![image](https://user-images.githubusercontent.com/746152/47939737-165f4280-dec7-11e8-95b7-080d5d6c4af8.png)

jeffersonrabb commented:
> Seems like Mapbox refers to the secret as Access tokens instead of keys.

Holding on this until we determine which service we're using, but if we land on Mapbox I'll make sure this text change happens.

simison commented:
Perhaps this PR could be merged and iterated further in smaller follow up PRs?

It can be left out of the preset-files before we publish to NPM in case you don't want to include in Jetpack plugin yet.

Just needs a rebase first and two tests are failing on Circle CI.

roccotripaldi commented on client/gutenberg/extensions/presets/jetpack/index.json:
```diff
@@ -1,5 +1,6 @@
 [
   "markdown",
   "publicize",
+  "map",
```
Let's get this removed from the preset and rebased.

jeffersonrabb commented:
> I assume this can now be closed, since the block was merged in #28498?

Correct. I'm closing now.
